### DISCUSSION TITLE
Расчёт суммы строк с учётом фильтров по дате.

### DISCUSSION THREAD

**Discussion by Ekhorikov2** [totumonline/totum-ru-issues-and-discussions#216] [Как это сделать?] [answered]
Всем привет. HELP.
Есть ПРОСТАЯ таблица, для учёта входящих и исходящих платежей.
![image](https://user-images.githubusercontent.com/112487577/188700942-fe9c4af8-258d-4e0d-9a92-d3932b76601f.png)
Для полей Приход и Расход, через Хедер, настроен расчёт суммы, через setFormat. Расчёт суммы производится с учётом всех фильтров, кроме Дата от и Дата до. Т.е. при выборе любого из фильтров, сумма пересчитывается с учётом отфильтрованных данных.
 ![image](https://user-images.githubusercontent.com/112487577/188701401-4da1ebed-400d-4749-8c72-c970850cf8a9.png)
Проблема в том, что если добавлять префильтры Дата от и Дата до в selectList, расчёт суммы перестаёт работать, значение отображается 0!
Всё победил, по всем фильтрам работает, по префильтру при открытии списка платежей (через iframe) с привязкой к проекту, тоже считает сумму только отображаемых в iframe платежей, а вот с датой не могу сообразить как быть. Значение 0!
Подскажите, как настроить чтоб при выборе диапазона дат, считалась сумму отображаемых строк?

**Comment by erpic-ru** [comment, marked as answer]
Должно работать так: `where 'data' >= #fl_data_ot; where 'data' <= #fl_data_do;` 

Но тут ещё на обязательность указания дат в префильтре можно посмотреть. Если заполнена только одна дата или не заполнены обе, то вкрутить какой-нибудь костыль типа:

`date1: if(condition: #fl_data_ot = ""; then: "1970-01-01"; else: #fl_data_ot)` и уже date1 в selectlist брать. 
Либо дефолтные значения для префильтров по датам выставить.

Наверное, получше вариант есть

**Reply by ttm-alexeypolunin** [reply]
Если передать `*ALL*` в параметр `where`, то он отключается:

```
 ...; where 'data' >= $from; where 'data' <= $to; ...

from: if(condition: #fl_data_ot = ""; the: "*ALL*"; else: #fl_data_ot)
to: if(condition: #fl_data_do = ""; the: "*ALL*"; else: #fl_data_do)
```

В основном в `дата от` — `дата до` ошибаются в направлении больше/меньше. Выбираемые значения должны быть:

— **больше** чем `ОТ`
— **меньше** чем `ДО`.

**Reply by Ekhorikov2** [reply]
Всё получилось. Спасибо огромное.

**Reply by Ekhorikov2** [reply]
Столкнулся ещё с такой проблемой - если поставить параметр "Множественные значения" у префильтра (для фильтрации по нескольким значениям), сумма тоже перестаёт считаться, выдаёт 0.

**Reply by ttm-alexeypolunin** [reply]
**Some table:**

<img width="580" alt="Снимок экрана 2022-09-14 в 17 18 13" src="https://user-images.githubusercontent.com/55755565/190179936-5640156d-3338-43ad-8213-a50cdb175aea.png">

**Code format:**

<img width="862" alt="Снимок экрана 2022-09-14 в 17 18 27" src="https://user-images.githubusercontent.com/55755565/190180063-f40f519d-43b1-47af-9df5-cc204f834a18.png">

```
f1=: setFormat(text: $sum)

	sum: listSum(list: $list)
		list: if(condition: #fl_prefilter = "*ALL*"; then: $all; else: $filter)
			filter: selectList(table: $#ntn; field: 'money'; where: 'id' = #fl_prefilter)
			all: selectList(table: $#ntn; field: 'money')
```

**Reply by Ekhorikov2** [reply]
А если несколько фильтров, на несколько полей? Тогда условие пересекаться будет? Я имею в виду, если нужно чтоб одновременно два или три фильтра работали.

**Comment by Ekhorikov2** [comment]
У меня сейчас вот что получается
![image](https://user-images.githubusercontent.com/112487577/190991300-cc4916a4-bd0f-450f-9e3e-d5ac578b908c.png)
Секция f1, срабатывает в том случае, если в технический префильтр (он сделан отдельно и скрыт от пользователей, для вывода привязанных к проекту платежей в iframe, чтоб не пересекался с другими префильтрами в кодах) передаётся значение проекта. Расчёт суммы производится с учётом только технического префильтра.
Секция f2, если технический префильтр пуст, тогда производится расчёт по всем остальным префильтрам, включая префильтры по датам и всё работает.
Если добавить вышеуказанный код для расчёта с множественными значениями, тоже работает, но только с каким-то одним префильтром, а у меня их в таблице 5 (которые могут быть множественными. Если повторять секции к разным префильтрам, они начинают друг другу противоречить, в результате чего срабатывает (естественно) опять только один.
